pennylane/__init__.py

```diff
@@ -337,3 +338,16 @@ def run_cnot():
 def version():
     """Returns the PennyLane version number."""
     return __version__
+
+
+def __getattr__(name):
+    """Re-route calls of qml.grouping as it will be deprecated."""
+    if name == "grouping":
+        warn(
+            "The qml.grouping module is deprecated, please use qml.pauli instead.",
+            DeprecationWarning,
+        )
+        from pennylane import grouping  # pylint:disable=import-outside-toplevel
+
+        return grouping
+    raise AttributeError(f"Module {__name__} has no attribute {name}")
```

Comment:
@Jaybsoni wait, this makes me really nervous - the logic seems to be 'if name is not grouping, raise error'!

Wouldn't this cause everything in PennyLane to raise an error?

Reply:
Also, this would appear to raise an error on every import? Not just the first import?

Reply:
@josh146, 

> Also, this would appear to raise an error on every import? Not just the first import?

Yes! This will appear on every import, but that is actually what we want. 

> @Jaybsoni wait, this makes me really nervous - the logic seems to be 'if name is not grouping, raise error'!
> 
> Wouldn't this cause everything in PennyLane that _isn't_ in the grouping module to raise an error?

Actually it doesn't raise such an error. My current theory is that the `__getattr__` method is only called after looking in the namespace of the file. Since everything else was imported into the name space earlier, we don't have to worry about raising errors for known modules in PennyLane. This only gets called when an attribute not imported or defined above is accessed.